### PR TITLE
zsjed: init at 0.3

### PR DESCRIPTION
**Pull request by fogti**
###### Motivation for this change

Packaging my semi-stenography tool (hiding file snippets at the end of other files).

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux) (enabled by default)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -10686,6 +10686,16 @@
     github = "pulsation";
     githubId = 1838397;
   };
+  zseri = {
```
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:10689]: Please put this into a separate commit with the message ``maintainers: add zseri``.

File: pkgs/tools/misc/zsjed/default.nix [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+  postFixup = let
+    runtimePath = lib.makeBinPath buildInputs;
+  in
+  ''
+    sed -i "2 i export PATH=${runtimePath}:\$PATH" $out/bin/zsjed-mkls
+  '';
+
+  meta = with lib; {
+    # https://github.com/zserik/portage-zscheile/blob/ea553888fff7f9728d60082e224fd808f2a8924d/media-metad/zsjed/zsjed-0.3.ebuild
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/zsjed/default.nix:49]:
```suggestion
```

File: pkgs/tools/misc/zsjed/default.nix [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    description = "Zscheile data split semi-steganography tool";
+    # https://github.com/zserik/portage-zscheile/blob/ea553888fff7f9728d60082e224fd808f2a8924d/media-metad/zsjed/metadata.xml
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/zsjed/default.nix:51]:
```suggestion
```

File: pkgs/tools/misc/zsjed/default.nix [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    longDescription = ''
+      zsjed is a data split semi-steganography tool.
+      It splits a file and pastes these parts
+      at the end of many other files, recoverable.
+    '';
+    # https://github.com/zserik/portage-zscheile/blob/ea553888fff7f9728d60082e224fd808f2a8924d/eclass/zs-minimal.eclass
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/zsjed/default.nix:57]:
```suggestion
```

File: pkgs/tools/misc/zsjed/default.nix [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    mkdir -p $out/bin
+    install -t $out/bin zsjed-collect zsjed-clear zsjed-spread zsjed-mkls
+  '';
+
+  postFixup = let
+    runtimePath = lib.makeBinPath buildInputs;
+  in
+  ''
+    sed -i "2 i export PATH=${runtimePath}:\$PATH" $out/bin/zsjed-mkls
+  '';
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/zsjed/default.nix:46]:
```suggestion
  postFixup = ''
    sed -i "2 i export PATH=${lib.makeBinPath buildInputs}:\$PATH" $out/bin/zsjed-mkls
  '';
```

File: pkgs/tools/misc/zsjed/default.nix [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  ];
+
+  # can't check because all tools return retcode 1 by default
+  doCheck = false;
+  enableParallelBuilding = true;
+
+  # the Makefile does not work for this, as it assumes the wrong prefix
+  installPhase = ''
+    mkdir -p $out/bin
+    install -t $out/bin zsjed-collect zsjed-clear zsjed-spread zsjed-mkls
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/zsjed/default.nix:38]:
```suggestion
    install -tD $out/bin zsjed-collect zsjed-clear zsjed-spread zsjed-mkls
```

File: pkgs/tools/misc/zsjed/default.nix [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+  ];
+
+  # runtime deps of zsjed-mkls
+  buildInputs = [
+    coreutils # sort
+    findutils
+    gawk
+  ];
+
+  # can't check because all tools return retcode 1 by default
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/zsjed/default.nix:31]: We could work around this. Do you have tests?

File: pkgs/tools/misc/zsjed/default.nix [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+  src = fetchurl {
+    url = "https://ytrizja.de/distfiles/zsjed-${version}.tar.gz";
+    sha256 = "1rcrvrqjsv80ldz7cscqbrf98jb0yd2in97irrzlpask699z5k29";
+  };
+
+  nativeBuildInputs = [
+    coreutils # mkdir, install
+    gnused
+  ];
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/zsjed/default.nix:22]:
```suggestion
```
Should be included in stdenv.

File: pkgs/tools/misc/zsjed/default.nix [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+  # runtime deps of zsjed-mkls
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/zsjed/default.nix:24]:
```suggestion
```

File: pkgs/tools/misc/zsjed/default.nix (hunk 6 above)
Comment (zserik) [reply pkgs/tools/misc/zsjed/default.nix:38]: This doesn't work:
```
installing
install: failed to access 'D': No such file or directory
```

Comment (SuperSandro2000) [reply pkgs/tools/misc/zsjed/default.nix:38]: I think the D needs to go before the t.

File: pkgs/tools/misc/zsjed/default.nix (hunk 8 above)
Comment (zserik) [reply pkgs/tools/misc/zsjed/default.nix]: Is `diffutils` included in stdenv?

File: pkgs/tools/misc/zsjed/default.nix (hunk 7 above)
Comment (zserik) [reply pkgs/tools/misc/zsjed/default.nix]: I found a workaround. Is it ok that the check has some output, or should I forward everything to `/dev/null`?

File: pkgs/tools/misc/zsjed/default.nix [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+    rm -rf tmp
+  '';
+
+  # the Makefile does not work for this, as it assumes the wrong prefix
+  installPhase = ''
+    install -D -t $out/bin zsjed-collect zsjed-clear zsjed-spread zsjed-mkls
+  '';
+
+  postFixup = ''
+    sed -i "2 i export PATH=${lib.makeBinPath buildInputs}:\$PATH" $out/bin/zsjed-mkls
```
Comment (zserik) [inline pkgs/tools/misc/zsjed/default.nix:54]: This is suboptimal, but I'm unsure how to use the `substitute*` fn's to put the appropriate paths in there...

Comment (SuperSandro2000) [reply pkgs/tools/misc/zsjed/default.nix:54]: I think you might want to use makeWrapper here. You can take a look at somafm-cli for a simple example.

File: pkgs/tools/misc/zsjed/default.nix [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+    gawk
+  ];
+
+  doCheck = true;
+  enableParallelBuilding = true;
+
+  # doesn't test zsjed-mkls
+  checkPhase = ''
+    mkdir tmp
+    cd tmp
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/zsjed/default.nix:35]:
```suggestion
    tmp=$(mktemp -d)
    cd $tmp
```

File: pkgs/tools/misc/zsjed/default.nix [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+    touch f1 f2
+    echo f1 > l
+    echo f2 >> l
+    ../zsjed-clear l
+    echo "a b c d e f g h i j k" > i
+    ../zsjed-spread i l
+    ../zsjed-collect j l
+    cmp i j
+    cd ..
+    rm -rf tmp
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/zsjed/default.nix:45]:
```suggestion
```

File: pkgs/tools/misc/zsjed/default.nix (hunk 8 above)
Comment (zserik) [reply pkgs/tools/misc/zsjed/default.nix]: I assume for now that is it not in stdenv.

File: pkgs/tools/misc/zsjed/default.nix (hunk 12 above)
Comment (zserik) [reply pkgs/tools/misc/zsjed/default.nix]: Huh, does Nix automatically remove any temp folder created by `mktemp -d`?

File: pkgs/tools/misc/zsjed/default.nix [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+    url = "https://ytrizja.de/distfiles/zsjed-${version}.tar.gz";
+    sha256 = "1rcrvrqjsv80ldz7cscqbrf98jb0yd2in97irrzlpask699z5k29";
+  };
+
+  nativeBuildInputs = [
+    diffutils
+    makeWrapper
+  ];
+
+  buildInputs = [
```
Comment (zserik) [inline pkgs/tools/misc/zsjed/default.nix:25]: That `buildInputs` is handled equivalent to `depsHostTarget` is outright confusing... (see also: https://nixos.wiki/wiki/Cross_Compiling)

File: pkgs/tools/misc/zsjed/default.nix (hunk 12 above)
Comment (SuperSandro2000) [reply pkgs/tools/misc/zsjed/default.nix]: nix-collect-garbage deletes it.

File: pkgs/tools/misc/zsjed/default.nix [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+    coreutils # sort
+    findutils
+    gawk
+  ];
+
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/zsjed/default.nix:30]:
```suggestion
```

File: pkgs/tools/misc/zsjed/default.nix [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+    "$dst"/zsjed-collect j l
+    cmp i j
+    cd "$dst"
+    rm -rf "$tmp"
+  '';
+
+  # can't reuse the Makefile 'install' target, as it assumes the wrong prefix
+  installPhase = ''
+    install -D -t $out/bin zsjed-collect zsjed-clear zsjed-spread zsjed-mkls
+    wrapProgram $out/bin/zsjed-mkls --prefix PATH ":" "${lib.makeBinPath buildInputs}"
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/zsjed/default.nix:54]:
```suggestion
    wrapProgram $out/bin/zsjed-mkls --prefix PATH ":" "${lib.makeBinPath [ coreutils findutils gawk]}"
```

File: pkgs/tools/misc/zsjed/default.nix (hunk 13 above)
Comment (SuperSandro2000) [reply pkgs/tools/misc/zsjed/default.nix:25]: Rule of thumb: everything linked against goes to buildInputs and everything executed or what should be in PATH goes to nativeBuildInputs. About all other cross compile issues I would just wait until someone discovers problems while building.

File: pkgs/tools/misc/zsjed/default.nix (hunk 14 above)
Comment (zserik) [reply pkgs/tools/misc/zsjed/default.nix:30]: I guess, it's in stdenv, right?

Comment (zserik) [reply pkgs/tools/misc/zsjed/default.nix:30]: Is there a list of packages which are in stdenv?

**Comment by zserik** [comment]
Could I do anything to speed this process up?
(I think `nix.useSandbox` is enabled by default in NixOS, so that is already done; I'm unsure what "relevant documentation" in this case might be)

**Comment by SuperSandro2000** [comment]
> I think `nix.useSandbox` is enabled by default in NixOS, so that is already done

yes it is.

> I'm unsure what "relevant documentation" in this case might be

Not relevant for simple packages.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 112959` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>

  <summary>1 package failed to build and already failed to build on hydra master:</summary>
  <ul>
<li>zsjed: <a href="https://termbin.com/jy2o">plain log</a>  |  <a href="https://htmlpreview.github.io/?https://gist.githubusercontent.com/SuperSandroBot/68eca43f66cc1270fc6006b7dbe7bd00/raw/index.html">pretty log </a></li>
  </ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
zsjed:
<details>


`checkPhase` should probably contain `runHook preCheck` and `runHook postCheck`.

Near pkgs/tools/misc/zsjed/default.nix:29:3:

```
   |
29 |   checkPhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/tools/misc/zsjed/default.nix:46:3:

```
   |
46 |   installPhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
</details>
</details>

**Comment by zserik** [comment]
I'm probably going to port zsjed in a few weeks, and this program isn't critical, so I'm skipping this for now.
